File: src/soak-test.ts
import http from 'k6/http';
import { check, sleep } from 'k6';
import { Options } from 'k6/options';

// Soak Test Configuration - Extended duration to find memory leaks, degradation
export const options: Options = {
  stages: [
    { duration: '2m', target: 20 },   // Ramp up to moderate load
    { duration: '30m', target: 20 },  // Maintain for 30 minutes (extended test)
    { duration: '2m', target: 0 },    // Ramp down
  ],
  thresholds: {
    http_req_duration: ['p(95)<1000'],  // Should stay consistent
    http_req_failed: ['rate<0.05'],     // Low error tolerance
  },
};

const BASE_URL = 'https://jsonplaceholder.typicode.com';

interface Post {
  title: string;
  body: string;
  userId: number;
}

export default function (): void {
  // Simulate realistic user behavior over extended period

  // User browses posts
  let response = http.get(`${BASE_URL}/posts`);
  check(response, {
    'browse posts - status 200': (r) => r.status === 200,
  });
  sleep(2);

  // User reads a specific post
  const postId: number = Math.floor(Math.random() * 100) + 1;
  response = http.get(`${BASE_URL}/posts/${postId}`);
  check(response, {
    'read post - status 200': (r) => r.status === 200,
  });
  sleep(3);

  // User checks comments
  response = http.get(`${BASE_URL}/posts/${postId}/comments`);
  check(response, {
    'read comments - status 200': (r) => r.status === 200,
  });
  sleep(2);

  // User creates a new post (less common action)
  if (Math.random() < 0.3) { // 30% of users create content
    const payload: Post = {
      title: `Soak test post ${Date.now()}`,
      body: 'Long running test content',
      userId: Math.floor(Math.random() * 10) + 1,
    };

    response = http.post(`${BASE_URL}/posts`, JSON.stringify(payload), {
      headers: { 'Content-Type': 'application/json' },
    });

    check(response, {
      'create post - status 201': (r) => r.status === 201,
    });
    sleep(2);
  }

  // User browses other content
  response = http.get(`${BASE_URL}/users`);
  check(response, {
    'browse users - status 200': (r) => r.status === 200,
  });
  sleep(3);
}

File: grinder/scripts/soak_test.py
"""
Soak Test - Extended duration test for memory leaks and degradation
The Grinder Jython script
"""

from net.grinder.script.Grinder import grinder
from net.grinder.script import Test
from net.grinder.plugin.http import HTTPRequest
from HTTPClient import NVPair
import random

# Test configuration
test1 = Test(1, "GET /posts")
test2 = Test(2, "GET /posts/:id")
test3 = Test(3, "GET /posts/:id/comments")
test4 = Test(4, "POST /posts")

# Base URL
BASE_URL = "https://jsonplaceholder.typicode.com"

# HTTP client setup
request = HTTPRequest()

class TestRunner:
    """Test runner class for soak test"""

    def __call__(self):
        """Run soak test scenario"""

        # Test 1: GET all posts
        test1.record(request)
        request.GET(BASE_URL + "/posts")
        grinder.sleep(300)

        # Test 2: GET random post
        test2.record(request)
        post_id = random.randint(1, 100)
        request.GET(BASE_URL + "/posts/" + str(post_id))
        grinder.sleep(300)

        # Test 3: GET post comments
        test3.record(request)
        post_id = random.randint(1, 100)
        request.GET(BASE_URL + "/posts/" + str(post_id) + "/comments")
        grinder.sleep(300)

        # Test 4: POST create post
        test4.record(request)
        headers = [NVPair("Content-Type", "application/json")]
        user_id = random.randint(1, 10)
        body = '{"title":"Soak test ' + str(random.randint(1, 1000)) + '","body":"Testing for extended duration","userId":' + str(user_id) + '}'
        request.POST(BASE_URL + "/posts", body, headers)
        grinder.sleep(300)
